snuba/admin/audit_log/base.py

```diff
@@ -0,0 +1,36 @@
+from abc import ABC, abstractmethod
+from datetime import datetime
+from typing import Any, Mapping, Optional
+
+
+class AuditLog(ABC):
```

Comment:
If I understand this design correctly, I think this class has two independent responsibilities that are going to generate a lot of duplication.
On one side it is feature specific: it runs the application logic that formats the audit log record or notification payload.
On the other side it contains the channel specific logic that allows you to write the record somewhere.

The consequence of storing both in the same class is that if you want to allow the multiple audit log type to use the same channel (like writing on redis) you need to duplicate the channel specific logic.

Why not making the `AuditLog` class represent a channel instead (if you can define a common interface). At that point you will have one implementation for slack, one for the GCP logger, one for redis, etc. Each use case can use channels the way it wants without the need for an abstract class everyone needs to implement.
A sign that this class in itself is not very useful are:
- I cannot imagine a callsite referencing the abstract class `AuditLog`. Whenever you use the audit log you only instantiate the subclasses and reference the subclasses. So this class is not useful for polymorphism.
- It does almost anything that is shared by subclasses. Which means you are not using to share code. Also there are better ways to share code via composition if polymorphism is not your goal.

If you then find value in adding common logic between all the audit log implementations you can add an `audit` function that implements this logic and forwards the payload to the specific channel. An example could be making each feature load the right channel according to configuration.

Reply:
okay what about something like: https://github.com/getsentry/snuba/pull/3406/commits/163749b38804677d198fd41238f1d7944878f1a4
which just lets people override the `_format_data` if they want to. In the future I could see refactoring this to have the audit log take maybe a notifier and a recorder or something. In my mind you'd always want to record the audit log, but it would optional if you wanted to get a notification, which is why I wasn't treating slack and GCP as the same (like as a channel), although you could argue they do just send the data somewhere. Not sure if im making sense here.

Reply:
> lets people override the _format_data if they want to. In the future I could see refactoring this to have the audit log take maybe a notifier and a recorder or something.

I think the original concern still applies. Why do you need a parent class for all subscriptions? There generally are two reasons for it: polymorphism and sharing code. 
Sharing code is generally better achieved by composition (put the common logic in its own class/function and use it from the code that wants to send the notification). You leave more flexibility to the callsite instead of adding a constraint that is not providing value.

Polymorphism does not seem to be your goal here. It seems to me you are never going to have a function or class that depends on `AurditLog`. They will always depend on the concrete implementation.

Also how would you implement two types of logs (logger and redis) with thelast version you submitted ?